source/module_hsolver/diago_subspacedav.cpp

```diff
@@ -0,0 +1,976 @@
+                             this->dim,
+                             this->zero,
+                             scc + nbase,
+                             this->nbase_x);
+    }
+
+#ifdef __MPI
+    if (GlobalV::NPROC_IN_POOL > 1)
+    {
+        matrixTranspose_op<T, Device>()(this->ctx, this->nbase_x, this->nbase_x, hcc, hcc);
+        matrixTranspose_op<T, Device>()(this->ctx, this->nbase_x, this->nbase_x, scc, scc);
+
+        auto* swap = new T[notconv * this->nbase_x];
+        syncmem_complex_op()(this->ctx, this->ctx, swap, hcc + nbase * this->nbase_x, notconv * this->nbase_x);
+
+        if (std::is_same<T, double>::value)
+        {
+            Parallel_Reduce::reduce_pool(hcc + nbase * this->nbase_x, notconv * this->nbase_x);
+            Parallel_Reduce::reduce_pool(scc + nbase * this->nbase_x, notconv * this->nbase_x);
+        }
+        else
+        {
+            if (psi::device::get_current_precision(swap) == "single")
+            {
+                MPI_Reduce(swap,
+                           hcc + nbase * this->nbase_x,
+                           notconv * this->nbase_x,
+                           MPI_COMPLEX,
+                           MPI_SUM,
+                           0,
+                           POOL_WORLD);
+            }
+            else
+            {
+                MPI_Reduce(swap,
+                           hcc + nbase * this->nbase_x,
+                           notconv * this->nbase_x,
+                           MPI_DOUBLE_COMPLEX,
+                           MPI_SUM,
+                           0,
+                           POOL_WORLD);
+            }
+
+            syncmem_complex_op()(this->ctx, this->ctx, swap, scc + nbase * this->nbase_x, notconv * this->nbase_x);
+
+            if (psi::device::get_current_precision(swap) == "single")
+            {
+                MPI_Reduce(swap,
+                           scc + nbase * this->nbase_x,
+                           notconv * this->nbase_x,
+                           MPI_COMPLEX,
+                           MPI_SUM,
+                           0,
+                           POOL_WORLD);
+            }
+            else
+            {
+                MPI_Reduce(swap,
+                           scc + nbase * this->nbase_x,
+                           notconv * this->nbase_x,
+                           MPI_DOUBLE_COMPLEX,
+                           MPI_SUM,
+                           0,
+                           POOL_WORLD);
+            }
+        }
+        delete[] swap;
+
+        matrixTranspose_op<T, Device>()(this->ctx, this->nbase_x, this->nbase_x, hcc, hcc);
+        matrixTranspose_op<T, Device>()(this->ctx, this->nbase_x, this->nbase_x, scc, scc);
+    }
+#endif
+
+    nbase += notconv;
+    int nb1 = nbase - notconv;
+    // reset:
+    if (init)
+    {
+        for (size_t i = 0; i < nbase; i++)
+        {
+
+            hcc[i * this->nbase_x + i] = set_real_tocomplex(hcc[i * this->nbase_x + i]);
```

Comment:
this 'i * this->nbase_x + i' can be calculated only once and use a variable to store the result